### PR TITLE
JAPI-503 String processing JUnit failure

### PR DESCRIPTION
**Pull request by jpmcmu**
## Type of change:
- [X] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [ ] This change does not cause any existing JUnits to fail.
  - [ ] I have include JUnit coverage to test this change
  - [ ] I have performed system test and covered possible regressions and side effects.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] This change fixes the problem, not just the symptom

### CODE REVIEW

**Comment by jpmcmu** [comment]
@rpastrana please review

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java [hunk 1]
```diff
@@ -669,20 +669,22 @@ public void stringProcesingTests() throws Exception
...
+        fieldDefs[5] = new FieldDef("str6", FieldType.STRING, "FIXED_SBC", 27, true, false, HpccSrcType.SINGLE_BYTE_CHAR, new FieldDef[0]);
+        fieldDefs[6] = new FieldDef("str7", FieldType.STRING, "FIXED_UNCIODE", 27, true, false, HpccSrcType.UTF16LE, new FieldDef[0]);
+        fieldDefs[7] = new FieldDef("str8", FieldType.VAR_STRING, "FIXED_VAR_UNICODE", 27, true, false, HpccSrcType.UTF16LE, new FieldDef[0]);
+        fieldDefs[8] = new FieldDef("str9", FieldType.VAR_STRING, "FIXED_VAR_STRING", 27, true, false, HpccSrcType.SINGLE_BYTE_CHAR, new FieldDef[0]);
 
         FieldDef recordDef = new FieldDef("RootRecord", FieldType.RECORD, "rec", 4, false, false, HpccSrcType.LITTLE_ENDIAN, fieldDefs);
 
         List<HPCCRecord> records = new ArrayList<HPCCRecord>();
 
+        String whiteSpaceStr = " \t\n";
```
Comment (rpastrana) [inline dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:681]: Perhaps overkill, but \p{javaWhitespace} would account for all whitespaces recognized by java.
https://stackoverflow.com/questions/17140158/why-isnt-char-of-type-space-separator-recognized-as-whitespace

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java [hunk 2]
```diff
@@ -669,20 +669,22 @@ public void stringProcesingTests() throws Exception
         fieldDefs[2] = new FieldDef("str3", FieldType.STRING, "UNICODE", 4, false, false, HpccSrcType.UTF16LE, new FieldDef[0]);
         fieldDefs[3] = new FieldDef("str4", FieldType.VAR_STRING, "VAR_UNICODE", 4, false, false, HpccSrcType.UTF16LE, new FieldDef[0]);
         fieldDefs[4] = new FieldDef("str5", FieldType.VAR_STRING, "VAR_STRING", 4, false, false, HpccSrcType.SINGLE_BYTE_CHAR, new FieldDef[0]);
-        fieldDefs[5] = new FieldDef("str6", FieldType.STRING, "FIXED_SBC", 12, true, false, HpccSrcType.SINGLE_BYTE_CHAR, new FieldDef[0]);
-        fieldDefs[6] = new FieldDef("str7", FieldType.STRING, "FIXED_UNCIODE", 12, true, false, HpccSrcType.UTF16LE, new FieldDef[0]);
-        fieldDefs[7] = new FieldDef("str8", FieldType.VAR_STRING, "FIXED_VAR_UNICODE", 12, true, false, HpccSrcType.UTF16LE, new FieldDef[0]);
-        fieldDefs[8] = new FieldDef("str9", FieldType.VAR_STRING, "FIXED_VAR_STRING", 12, true, false, HpccSrcType.SINGLE_BYTE_CHAR, new FieldDef[0]);
+        fieldDefs[5] = new FieldDef("str6", FieldType.STRING, "FIXED_SBC", 27, true, false, HpccSrcType.SINGLE_BYTE_CHAR, new FieldDef[0]);
```
Comment (rpastrana) [inline dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:672]: what's causing the increase by 15

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java [hunk 3]
```diff
@@ -965,7 +965,7 @@ private void trimStringInScratchBuffer(boolean isUnicode, int[] range)
             while (range[1] > range[0])
             {
                 // Need to check against EOS (0x0) in trim fixed len strings correctly
-                int codePoint = this.scratchBuffer[range[1]-2] << 8 | this.scratchBuffer[range[1]-1];
+                int codePoint = (this.scratchBuffer[range[1]-2] & 0xFF) | ((this.scratchBuffer[range[1]-1] & 0xFF) << 8);
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:968]: what is the intended net effect of the bitwise and operation w/ the FF mask?

**Review by rpastrana — COMMENTED** [review]
@jpmcmu a couple of questions

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java (hunk 2 above)
Comment (jpmcmu) [reply dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:672]: The size increased due to the additional white space, but I also increased it further to an odd number to make sure alignment issues weren't also a problem.

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java (hunk 3 above)
Comment (jpmcmu) [reply dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:968]: Idea here is to combine those two bytes into an unsigned short for the unicode code point. Masking by 0xFF to convert from signed byte to unsigned value in an int, shift upper byte by 8 bits, and then OR the lower and upper ints together to get desired unsigned code point.

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java (hunk 1 above)
Comment (jpmcmu) [reply dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:681]: Oh, that is cool! I will do that

Comment (jpmcmu) [reply dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:681]: Hmm, looks like that escape only works inside of regex?

Comment (rpastrana) [reply dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:681]: did you look for other (non regex) ways to accomplish the same? BTW I didn't imply testing for all known Java whitespaces make sense.

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java (hunk 2 above)
Comment (rpastrana) [reply dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:672]: k, that wasn't obvious from code to the untrained eye, let's comment any "odd" scenario like this one for the sake of maintenance.

**Comment by jpmcmu** [comment]
@rpastrana please review

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java [hunk 4]
```diff
@@ -953,7 +953,7 @@ private void trimStringInScratchBuffer(boolean isUnicode, int[] range)
         {
             while (range[0] < range[1] - 1)
             {
-                int codePoint = this.scratchBuffer[range[0]] << 8 | this.scratchBuffer[range[0]+1];
+                int codePoint = (this.scratchBuffer[range[0]] & 0xFF) | ((this.scratchBuffer[range[0]+1] & 0xFF) << 8);
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:956]: @jpmcmu this needs a source code comment, let's make sure the next person working on this understands why this is being done.

**Review by rpastrana — APPROVED** [review]

**Pull request merged**